{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "derived-secretariat",
   "metadata": {},
   "source": [
    "# Analysis of long-read transcriptomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "built-tournament",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import anndata\n",
    "import scipy.stats as st\n",
    "import statsmodels.stats as stm\n",
    "from statsmodels.stats.multitest import multipletests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "helpful-perspective",
   "metadata": {},
   "source": [
    "### Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "streaming-nirvana",
   "metadata": {},
   "source": [
    "Run Minimap2, TranscriptClean, and TALON using `run_talon_tc.sh`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baking-austria",
   "metadata": {},
   "source": [
    "Create TALON abundance file\n",
    "```bash \n",
    "db=talon/pgp1.db\n",
    "talon_abundance \\\n",
    "    --db $db \\\n",
    "    -a gencode_v29 \\\n",
    "    -b hg38 \\\n",
    "    --o talon/pgp1\n",
    "```\n",
    "\n",
    "Filter novel transcripts for reproducibility\n",
    "```bash\n",
    "db=talon/pgp1.db\n",
    "talon_filter_transcripts \\\n",
    "    --db $db \\\n",
    "    -a gencode_v29 \\\n",
    "    --maxFracA=0.5 \\\n",
    "    --minCount=5 \\\n",
    "    --minDatasets=2 \\\n",
    "    --o talon/pgp1_pass_list.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "knowing-drinking",
   "metadata": {},
   "source": [
    "### TSS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "rocky-playing",
   "metadata": {},
   "source": [
    "Isolate reads that represent more confident 5' ends."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "undefined-control",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/anndata/_core/anndata.py:21: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import RangeIndex\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import swan_vis as swan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "proud-staff",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot = 'talon/pgp1_talon_read_annot.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "collective-ethnic",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit the reads to those that represent putative 5' ends\n",
    "df = pd.read_csv(annot, sep='\\t')\n",
    "tss_df = df.loc[df.transcript_novelty.isin(['Known', 'NIC', 'NNC', 'ISM'])]\n",
    "tss_df = tss_df.loc[tss_df.ISM_subtype.isin(['None', 'Prefix', 'Both'])]\n",
    "tss_reads = tss_df.read_name.tolist()\n",
    "\n",
    "# tss\n",
    "fname = 'tss_read_names.txt'\n",
    "with open(fname, 'w') as ofile:\n",
    "    for r in tss_reads:\n",
    "        ofile.write(r+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "important-costume",
   "metadata": {},
   "source": [
    "Create a sam file with all the TSS reads from the merged BAM using picard tools `isolate_tss_reads.sh`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "separate-belarus",
   "metadata": {},
   "source": [
    "Call TSSs using Diane's script\n",
    "```bash\n",
    "tss_dir=~/mortazavi_lab/bin/tss-annotation/long_read/\n",
    "python ${tss_dir}pacbio_to_tss.py \\\n",
    "    -i tss_reads.bam \\\n",
    "    --window-size=50 \\\n",
    "    --expression-threshold=2 \\\n",
    "    -o unfilt_tss.bed \\\n",
    "    -r \\\n",
    "    -n rev_tss.bw \\\n",
    "    -p fwd_tss.bw\n",
    "```\n",
    "\n",
    "Get the read names associated with each TSS as well \n",
    "```bash\n",
    "python ${tss_dir}tss_reads.py \\\n",
    "    -i talon_tmp/merged.bam \\\n",
    "    -r unfilt_tss.bed \\\n",
    "    -o tss_reads.bed\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "disturbed-equivalent",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/ipykernel_launcher.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  isetter(loc, value)\n"
     ]
    }
   ],
   "source": [
    "# filter a list of TSSs for each gene\n",
    "annot = 'talon/pgp1_talon_read_annot.tsv'\n",
    "tss_reads = 'tss_reads.bed'\n",
    "\n",
    "df = pd.read_csv(annot, sep='\\t')\n",
    "\n",
    "# remove sirvs and erccs\n",
    "df = df.loc[~df.chrom.str.contains('SIRV')]\n",
    "df = df.loc[~df.chrom.str.contains('ERCC')]\n",
    "\n",
    "# print('Before assigning each read a TSS')\n",
    "# print(len(df.index))\n",
    "ends = pd.read_csv(tss_reads, sep='\\t', header=None,\n",
    "            names=['chrom', 'start', 'end', 'read_name', 'tss_id', 'stand'])\n",
    "\n",
    "# merge on read name\n",
    "df = df.merge(ends, how='inner', on='read_name')\n",
    "\n",
    "# groupby on gene and tss\n",
    "df = df[['read_name', 'annot_gene_name', 'tss_id']].groupby(['annot_gene_name', 'tss_id']).count()\n",
    "\n",
    "# filter tsss for those that have >10% of the reads\n",
    "# for the most highly-expressed tss of the gene\n",
    "df.reset_index(inplace=True)\n",
    "df.rename({'read_name':'count'}, axis=1, inplace=True)\n",
    "df.loc[df.annot_gene_name == 'A1BG', 'count'].max()\n",
    "temp = df.loc[df.apply(lambda x: x['count'] >= df.loc[df.annot_gene_name==x.annot_gene_name, 'count'].max()*0.1, axis=1)]\n",
    "\n",
    "temp.loc[temp.annot_gene_name == 'MBP']\n",
    "\n",
    "#, assign each TSS a name, and quantify TSS exp\n",
    "temp.sort_values(by=['annot_gene_name'], inplace=True)\n",
    "temp['tss_id_2'] = np.nan\n",
    "prev_gene = None\n",
    "for ind, entry in temp.iterrows():\n",
    "    curr_gene = entry.annot_gene_name\n",
    "    if curr_gene != prev_gene:\n",
    "        i = 1\n",
    "    else:\n",
    "        i += 1\n",
    "    prev_gene = curr_gene\n",
    "    temp.loc[ind, 'tss_id_2'] = '{}_{}'.format(curr_gene, i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "strategic-garage",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>count</th>\n",
       "      <th>tss_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>m54284U_210303_010717/56821818/ccs_11277</td>\n",
       "      <td>70</td>\n",
       "      <td>A1BG_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>m54284_190620_121340/10027666/31_1705_CCS_11278</td>\n",
       "      <td>68</td>\n",
       "      <td>A1BG-AS1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>m54284_190620_121340/62456306/4696_57_CCS_4485</td>\n",
       "      <td>43</td>\n",
       "      <td>A2M_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>m54284_190620_121340/71565601/31_2054_CCS_14051</td>\n",
       "      <td>38</td>\n",
       "      <td>A4GALT_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>m54284_190612_161714/42402034/31_1873_CCS_4799</td>\n",
       "      <td>185</td>\n",
       "      <td>AAAS_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_gene_name                                           tss_id  count  \\\n",
       "0            A1BG         m54284U_210303_010717/56821818/ccs_11277     70   \n",
       "1        A1BG-AS1  m54284_190620_121340/10027666/31_1705_CCS_11278     68   \n",
       "2             A2M   m54284_190620_121340/62456306/4696_57_CCS_4485     43   \n",
       "4          A4GALT  m54284_190620_121340/71565601/31_2054_CCS_14051     38   \n",
       "5            AAAS   m54284_190612_161714/42402034/31_1873_CCS_4799    185   \n",
       "\n",
       "     tss_id_2  \n",
       "0      A1BG_1  \n",
       "1  A1BG-AS1_1  \n",
       "2       A2M_1  \n",
       "4    A4GALT_1  \n",
       "5      AAAS_1  "
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "id": "detected-secret",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>count</th>\n",
       "      <th>tss_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>m54284_190610_233512/51970467/31_711_CCS_16063</td>\n",
       "      <td>11</td>\n",
       "      <td>AADAT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>AASS</td>\n",
       "      <td>m54284_190613_123747/67633476/29_3042_CCS_19077</td>\n",
       "      <td>243</td>\n",
       "      <td>AASS_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ABAT</td>\n",
       "      <td>m54284U_210304_072206/27723898/ccs_7420</td>\n",
       "      <td>26</td>\n",
       "      <td>ABAT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ABCB6</td>\n",
       "      <td>m54284_190618_193207/46662214/30_2343_CCS_12604</td>\n",
       "      <td>29</td>\n",
       "      <td>ABCB6_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>ABCC10</td>\n",
       "      <td>m54284_190618_193207/47841528/31_2258_CCS_17733</td>\n",
       "      <td>16</td>\n",
       "      <td>ABCC10_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23319</th>\n",
       "      <td>ZSCAN12</td>\n",
       "      <td>m54284_190620_121340/73597637/3574_56_CCS_17396</td>\n",
       "      <td>13</td>\n",
       "      <td>ZSCAN12_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23330</th>\n",
       "      <td>ZSCAN30</td>\n",
       "      <td>m54284U_210303_010717/132186743/ccs_9616</td>\n",
       "      <td>15</td>\n",
       "      <td>ZSCAN30_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23333</th>\n",
       "      <td>ZSCAN31</td>\n",
       "      <td>m54284_190617_231034/11600701/42_1820_CCS_17393</td>\n",
       "      <td>4</td>\n",
       "      <td>ZSCAN31_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23338</th>\n",
       "      <td>ZSCAN5A</td>\n",
       "      <td>m54284_190619_155256/40764222/29_1368_CCS_11212</td>\n",
       "      <td>10</td>\n",
       "      <td>ZSCAN5A_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23342</th>\n",
       "      <td>ZSWIM7</td>\n",
       "      <td>m54284_190617_231034/34668949/30_2385_CCS_8441</td>\n",
       "      <td>47</td>\n",
       "      <td>ZSWIM7_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2821 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      annot_gene_name                                           tss_id  count  \\\n",
       "9               AADAT   m54284_190610_233512/51970467/31_711_CCS_16063     11   \n",
       "25               AASS  m54284_190613_123747/67633476/29_3042_CCS_19077    243   \n",
       "28               ABAT          m54284U_210304_072206/27723898/ccs_7420     26   \n",
       "35              ABCB6  m54284_190618_193207/46662214/30_2343_CCS_12604     29   \n",
       "41             ABCC10  m54284_190618_193207/47841528/31_2258_CCS_17733     16   \n",
       "...               ...                                              ...    ...   \n",
       "23319         ZSCAN12  m54284_190620_121340/73597637/3574_56_CCS_17396     13   \n",
       "23330         ZSCAN30         m54284U_210303_010717/132186743/ccs_9616     15   \n",
       "23333         ZSCAN31  m54284_190617_231034/11600701/42_1820_CCS_17393      4   \n",
       "23338         ZSCAN5A  m54284_190619_155256/40764222/29_1368_CCS_11212     10   \n",
       "23342          ZSWIM7   m54284_190617_231034/34668949/30_2385_CCS_8441     47   \n",
       "\n",
       "        tss_id_2  \n",
       "9        AADAT_2  \n",
       "25        AASS_2  \n",
       "28        ABAT_2  \n",
       "35       ABCB6_2  \n",
       "41      ABCC10_2  \n",
       "...          ...  \n",
       "23319  ZSCAN12_2  \n",
       "23330  ZSCAN30_2  \n",
       "23333  ZSCAN31_2  \n",
       "23338  ZSCAN5A_2  \n",
       "23342   ZSWIM7_2  \n",
       "\n",
       "[2821 rows x 4 columns]"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.tss_id_2.str.endswith('_2')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "literary-burden",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>count</th>\n",
       "      <th>tss_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [annot_gene_name, tss_id, count, tss_id_2]\n",
       "Index: []"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.tss_id_2.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "id": "derived-cancer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>count</th>\n",
       "      <th>tss_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>m54284U_210304_072206/171836912/ccs_16062</td>\n",
       "      <td>13</td>\n",
       "      <td>AADAT_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>m54284_190610_233512/51970467/31_711_CCS_16063</td>\n",
       "      <td>11</td>\n",
       "      <td>AADAT_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AADAT</td>\n",
       "      <td>m54284_190612_161714/13894049/30_2117_CCS_16061</td>\n",
       "      <td>2</td>\n",
       "      <td>AADAT_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   annot_gene_name                                           tss_id  count  \\\n",
       "8            AADAT        m54284U_210304_072206/171836912/ccs_16062     13   \n",
       "9            AADAT   m54284_190610_233512/51970467/31_711_CCS_16063     11   \n",
       "10           AADAT  m54284_190612_161714/13894049/30_2117_CCS_16061      2   \n",
       "\n",
       "   tss_id_2  \n",
       "8   AADAT_1  \n",
       "9   AADAT_2  \n",
       "10  AADAT_3  "
      ]
     },
     "execution_count": 329,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.annot_gene_name == 'AADAT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "verbal-ghost",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged called TSSs back in with read annot\n",
    "df = pd.read_csv(annot, sep='\\t')\n",
    "ends = pd.read_csv(tss_reads, sep='\\t', header=None,\n",
    "            names=['chrom', 'start', 'end', 'read_name', 'tss_id', 'stand'])\n",
    "ends = ends[['read_name', 'tss_id']]\n",
    "\n",
    "# merge on read name\n",
    "df = df.merge(ends, how='inner', on='read_name')\n",
    "temp = temp[['annot_gene_name', 'tss_id', 'tss_id_2']]\n",
    "df = df.merge(temp, how='inner', on=['annot_gene_name', 'tss_id'])\n",
    "\n",
    "# format like talon ab\n",
    "cols = ['annot_gene_name', 'annot_gene_id', 'dataset', 'tss_id_2']\n",
    "df = df[cols+['read_name']].groupby(cols).count().reset_index()\n",
    "df.rename({'read_name':'counts', 'tss_id_2':'tss_id'}, axis=1, inplace=True)\n",
    "df = df.pivot(index=['annot_gene_name', 'annot_gene_id', 'tss_id'], columns='dataset', values='counts')\n",
    "df.reset_index(inplace=True)\n",
    "df = df.rename_axis(None, axis=1)\n",
    "df.fillna(0, inplace=True)\n",
    "\n",
    "df.to_csv('pgp1_tss_talon_abundance.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "sticky-generation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>astro_1</th>\n",
       "      <th>astro_2</th>\n",
       "      <th>excite_neuron_1</th>\n",
       "      <th>excite_neuron_2</th>\n",
       "      <th>pgp1_1</th>\n",
       "      <th>pgp1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>A1BG_1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895.5</td>\n",
       "      <td>A1BG-AS1_1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899.14</td>\n",
       "      <td>A2M_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274.16</td>\n",
       "      <td>A4GALT_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>ENSG00000094914.12</td>\n",
       "      <td>AAAS_1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>61.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_gene_name       annot_gene_id      tss_id  astro_1  astro_2  \\\n",
       "0            A1BG  ENSG00000121410.11      A1BG_1      8.0      3.0   \n",
       "1        A1BG-AS1   ENSG00000268895.5  A1BG-AS1_1      8.0      3.0   \n",
       "2             A2M  ENSG00000175899.14       A2M_1      2.0      1.0   \n",
       "3          A4GALT  ENSG00000128274.16    A4GALT_1      1.0      0.0   \n",
       "4            AAAS  ENSG00000094914.12      AAAS_1     16.0      6.0   \n",
       "\n",
       "   excite_neuron_1  excite_neuron_2  pgp1_1  pgp1_2  \n",
       "0             18.0             11.0    22.0     8.0  \n",
       "1             26.0             23.0     5.0     3.0  \n",
       "2             11.0             29.0     0.0     0.0  \n",
       "3             19.0             15.0     2.0     1.0  \n",
       "4             20.0             29.0    53.0    61.0  "
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "grateful-psychology",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>astro_1</th>\n",
       "      <th>astro_2</th>\n",
       "      <th>excite_neuron_1</th>\n",
       "      <th>excite_neuron_2</th>\n",
       "      <th>pgp1_1</th>\n",
       "      <th>pgp1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4626</th>\n",
       "      <td>DIABLO</td>\n",
       "      <td>ENSG00000184047.17</td>\n",
       "      <td>DIABLO_1</td>\n",
       "      <td>63.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4627</th>\n",
       "      <td>DIABLO</td>\n",
       "      <td>ENSG00000284934.1</td>\n",
       "      <td>DIABLO_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4753</th>\n",
       "      <td>DNAJC9-AS1</td>\n",
       "      <td>ENSG00000227540.1</td>\n",
       "      <td>DNAJC9-AS1_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4754</th>\n",
       "      <td>DNAJC9-AS1</td>\n",
       "      <td>ENSG00000236756.4</td>\n",
       "      <td>DNAJC9-AS1_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7536</th>\n",
       "      <td>HSPA14</td>\n",
       "      <td>ENSG00000187522.15</td>\n",
       "      <td>HSPA14_1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7537</th>\n",
       "      <td>HSPA14</td>\n",
       "      <td>ENSG00000284024.2</td>\n",
       "      <td>HSPA14_1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10737</th>\n",
       "      <td>None</td>\n",
       "      <td>SIRV1.1</td>\n",
       "      <td>None_6</td>\n",
       "      <td>67.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10738</th>\n",
       "      <td>None</td>\n",
       "      <td>SIRV1.2</td>\n",
       "      <td>None_6</td>\n",
       "      <td>94.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10739</th>\n",
       "      <td>None</td>\n",
       "      <td>SIRV3.1</td>\n",
       "      <td>None_14</td>\n",
       "      <td>170.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10740</th>\n",
       "      <td>None</td>\n",
       "      <td>SIRV3.2</td>\n",
       "      <td>None_14</td>\n",
       "      <td>60.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11522</th>\n",
       "      <td>PINX1</td>\n",
       "      <td>ENSG00000254093.8</td>\n",
       "      <td>PINX1_1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11523</th>\n",
       "      <td>PINX1</td>\n",
       "      <td>ENSG00000258724.1</td>\n",
       "      <td>PINX1_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11555</th>\n",
       "      <td>PKD1P1</td>\n",
       "      <td>ENSG00000183889.12</td>\n",
       "      <td>PKD1P1_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11557</th>\n",
       "      <td>PKD1P1</td>\n",
       "      <td>ENSG00000244257.5</td>\n",
       "      <td>PKD1P1_2</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11798</th>\n",
       "      <td>POLR2J3</td>\n",
       "      <td>ENSG00000168255.20</td>\n",
       "      <td>POLR2J3_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11799</th>\n",
       "      <td>POLR2J3</td>\n",
       "      <td>ENSG00000285437.1</td>\n",
       "      <td>POLR2J3_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11800</th>\n",
       "      <td>POLR2J4</td>\n",
       "      <td>ENSG00000214783.9</td>\n",
       "      <td>POLR2J4_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11801</th>\n",
       "      <td>POLR2J4</td>\n",
       "      <td>ENSG00000272655.2</td>\n",
       "      <td>POLR2J4_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12921</th>\n",
       "      <td>RGS5</td>\n",
       "      <td>ENSG00000143248.12</td>\n",
       "      <td>RGS5_2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12922</th>\n",
       "      <td>RGS5</td>\n",
       "      <td>ENSG00000232995.7</td>\n",
       "      <td>RGS5_2</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13645</th>\n",
       "      <td>SCO2</td>\n",
       "      <td>ENSG00000130489.14</td>\n",
       "      <td>SCO2_1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13646</th>\n",
       "      <td>SCO2</td>\n",
       "      <td>ENSG00000284194.1</td>\n",
       "      <td>SCO2_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14638</th>\n",
       "      <td>SOD2</td>\n",
       "      <td>ENSG00000112096.17</td>\n",
       "      <td>SOD2_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14640</th>\n",
       "      <td>SOD2</td>\n",
       "      <td>ENSG00000285441.1</td>\n",
       "      <td>SOD2_1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15332</th>\n",
       "      <td>TBCE</td>\n",
       "      <td>ENSG00000284770.1</td>\n",
       "      <td>TBCE_1</td>\n",
       "      <td>30.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15333</th>\n",
       "      <td>TBCE</td>\n",
       "      <td>ENSG00000285053.1</td>\n",
       "      <td>TBCE_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      annot_gene_name       annot_gene_id        tss_id  astro_1  astro_2  \\\n",
       "4626           DIABLO  ENSG00000184047.17      DIABLO_1     63.0     31.0   \n",
       "4627           DIABLO   ENSG00000284934.1      DIABLO_1      1.0      0.0   \n",
       "4753       DNAJC9-AS1   ENSG00000227540.1  DNAJC9-AS1_1      0.0      0.0   \n",
       "4754       DNAJC9-AS1   ENSG00000236756.4  DNAJC9-AS1_1      0.0      0.0   \n",
       "7536           HSPA14  ENSG00000187522.15      HSPA14_1     17.0      9.0   \n",
       "7537           HSPA14   ENSG00000284024.2      HSPA14_1      9.0     10.0   \n",
       "10737            None             SIRV1.1        None_6     67.0     22.0   \n",
       "10738            None             SIRV1.2        None_6     94.0     49.0   \n",
       "10739            None             SIRV3.1       None_14    170.0    111.0   \n",
       "10740            None             SIRV3.2       None_14     60.0     18.0   \n",
       "11522           PINX1   ENSG00000254093.8       PINX1_1      6.0      5.0   \n",
       "11523           PINX1   ENSG00000258724.1       PINX1_1      0.0      1.0   \n",
       "11555          PKD1P1  ENSG00000183889.12      PKD1P1_2      2.0      7.0   \n",
       "11557          PKD1P1   ENSG00000244257.5      PKD1P1_2      4.0      2.0   \n",
       "11798         POLR2J3  ENSG00000168255.20     POLR2J3_1      2.0      2.0   \n",
       "11799         POLR2J3   ENSG00000285437.1     POLR2J3_1      3.0      5.0   \n",
       "11800         POLR2J4   ENSG00000214783.9     POLR2J4_1      0.0      2.0   \n",
       "11801         POLR2J4   ENSG00000272655.2     POLR2J4_1      2.0      0.0   \n",
       "12921            RGS5  ENSG00000143248.12        RGS5_2      2.0      0.0   \n",
       "12922            RGS5   ENSG00000232995.7        RGS5_2      3.0      0.0   \n",
       "13645            SCO2  ENSG00000130489.14        SCO2_1     17.0      5.0   \n",
       "13646            SCO2   ENSG00000284194.1        SCO2_1      0.0      0.0   \n",
       "14638            SOD2  ENSG00000112096.17        SOD2_1      0.0      0.0   \n",
       "14640            SOD2   ENSG00000285441.1        SOD2_1      3.0      1.0   \n",
       "15332            TBCE   ENSG00000284770.1        TBCE_1     30.0     12.0   \n",
       "15333            TBCE   ENSG00000285053.1        TBCE_1      0.0      0.0   \n",
       "\n",
       "       excite_neuron_1  excite_neuron_2  pgp1_1  pgp1_2  \n",
       "4626              68.0             85.0   106.0    79.0  \n",
       "4627               2.0              0.0     3.0     1.0  \n",
       "4753               4.0              1.0     5.0     0.0  \n",
       "4754               1.0              3.0     0.0     2.0  \n",
       "7536              13.0             13.0    67.0    49.0  \n",
       "7537               2.0              4.0     2.0     5.0  \n",
       "10737              0.0              0.0     0.0     0.0  \n",
       "10738              0.0              0.0     0.0     0.0  \n",
       "10739              0.0              0.0     0.0     0.0  \n",
       "10740              0.0              0.0     0.0     0.0  \n",
       "11522             13.0              8.0    40.0    22.0  \n",
       "11523              0.0              0.0     0.0     0.0  \n",
       "11555              0.0              0.0     0.0     0.0  \n",
       "11557              0.0              0.0     0.0     0.0  \n",
       "11798              7.0              6.0     5.0     2.0  \n",
       "11799             13.0             13.0     6.0     9.0  \n",
       "11800              5.0              6.0     4.0     3.0  \n",
       "11801              7.0              4.0     2.0     1.0  \n",
       "12921              0.0              0.0     2.0     1.0  \n",
       "12922              0.0              2.0     1.0     3.0  \n",
       "13645             37.0             30.0    32.0    16.0  \n",
       "13646              0.0              1.0     0.0     0.0  \n",
       "14638              0.0              1.0     1.0     0.0  \n",
       "14640              5.0              5.0     3.0     3.0  \n",
       "15332             21.0             26.0    29.0    32.0  \n",
       "15333              0.0              1.0     0.0     1.0  "
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(df.loc[df.tss_id.duplicated(keep=False)].index))\n",
    "df.loc[df.tss_id.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vanilla-median",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "graphic-reminder",
   "metadata": {},
   "source": [
    "### Swan"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "collected-arrival",
   "metadata": {},
   "source": [
    "Swan config file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "external-corpus",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Adding dataset annotation to the SwanGraph\n",
      "\n",
      "Adding dataset astro_1 to the SwanGraph\n",
      "\n",
      "Adding dataset astro_2 to the SwanGraph\n",
      "\n",
      "Adding dataset excite_neuron_1 to the SwanGraph\n",
      "\n",
      "Adding dataset excite_neuron_2 to the SwanGraph\n",
      "\n",
      "Adding dataset pgp1_1 to the SwanGraph\n",
      "\n",
      "Adding dataset pgp1_2 to the SwanGraph\n",
      "Saving graph as swan.p\n"
     ]
    }
   ],
   "source": [
    "annot = '/Users/fairliereese/mortazavi_lab/ref/gencode.v29/gencode.v29.SIRV.ERCC.annotation.gtf'\n",
    "\n",
    "# initialize SwanGraph\n",
    "sg = swan.SwanGraph()\n",
    "\n",
    "# add annotation GTF for reference\n",
    "sg.add_annotation(annot)\n",
    "\n",
    "# add transcript models from each dataset\n",
    "sg.add_datasets('swan_config.tsv')\n",
    "\n",
    "sg.save_graph('swan')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informational-flour",
   "metadata": {},
   "source": [
    "### TSS switching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "global-preliminary",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>A1BG_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895.5</td>\n",
       "      <td>A1BG-AS1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>ENSG00000268895.5</td>\n",
       "      <td>AC012313.2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENSG00000175899.14</td>\n",
       "      <td>A2M_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENSG00000128274.16</td>\n",
       "      <td>A4GALT_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name             gene_id        tss_id\n",
       "0      A1BG  ENSG00000121410.11        A1BG_1\n",
       "1  A1BG-AS1   ENSG00000268895.5    A1BG-AS1_1\n",
       "2  A1BG-AS1   ENSG00000268895.5  AC012313.2_1\n",
       "3       A2M  ENSG00000175899.14         A2M_1\n",
       "4    A4GALT  ENSG00000128274.16      A4GALT_1"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 308,
   "id": "postal-arabic",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df: talon abundance file (either tss or transcript)\n",
    "# cond_map: dictionary of {condition: [dataset1, dataset2]}; how you want to group datasets\n",
    "# how: whether to make a tss or iso level adata; 'tss' or 'iso'\n",
    "# pass_list: if 'iso', file of valid transcript IDs that pass filtering\n",
    "def make_adata(df, cond_map, how='iso', pass_list=None):\n",
    "    \n",
    "    # filter talon ab file based on pass list\n",
    "#     df = pd.read_csv(ab_file, sep='\\t')\n",
    "    if pass_list:\n",
    "        pass_list = pd.read_csv(pass_list, header=None, names=['gene_id', 'transcript_id'])\n",
    "        df = df.loc[df.transcript_ID.isin(pass_list.transcript_id.tolist())]\n",
    "\n",
    "    # obs table\n",
    "    obs = pd.DataFrame.from_dict(cond_map, orient='index')\n",
    "    obs.reset_index(inplace=True)\n",
    "    id_vars = ['index']\n",
    "    value_vars = obs.columns[1:]\n",
    "    obs = obs.melt(id_vars=id_vars, value_vars=value_vars)\n",
    "    obs.drop('variable', axis=1, inplace=True)\n",
    "    obs.rename({'index':'condition', 'value':'dataset'}, axis=1, inplace=True)\n",
    "\n",
    "    # var table\n",
    "    if how=='iso':\n",
    "        var_cols = ['annot_transcript_name', 'annot_gene_name',\\\n",
    "                    'annot_transcript_id', 'annot_gene_id', \\\n",
    "                    'gene_ID', 'transcript_ID', 'transcript_novelty', \\\n",
    "                    'ISM_subtype']\n",
    "        var = df[var_cols]\n",
    "        var.rename({'transcript_ID':'transcript_id', \\\n",
    "                    'gene_ID':'gene_id',\\\n",
    "                    'annot_gene_name': 'gene_name'}, axis=1, inplace=True)\n",
    "    if how=='tss': \n",
    "        var_cols = ['annot_gene_name', 'annot_gene_id', 'tss_id']\n",
    "        var = df[var_cols]\n",
    "        var.rename({'annot_gene_name': 'gene_name',\\\n",
    "                    'annot_gene_id':'gene_id'}, axis=1, inplace=True)\n",
    "        \n",
    "    # X table\n",
    "    df = df.transpose()\n",
    "    df = df.loc[df.index.isin(obs.dataset.tolist())]\n",
    "    obs_order = obs['dataset'].reset_index().set_index('dataset')\n",
    "    df['dataset_num'] = df.index.map(obs_order['index'])\n",
    "    df.sort_values('dataset_num', inplace=True)\n",
    "    df.drop('dataset_num', axis=1, inplace=True)\n",
    "    X = df.to_numpy()\n",
    "\n",
    "    adata = anndata.AnnData(obs=obs, var=var, X=X) \n",
    "    \n",
    "    return adata\n",
    "\n",
    "# gene_df: pandas dataframe with expression values in each condition for each TSS in a gene\n",
    "# conditions: list of str of condition names\n",
    "# rc: threshold of read count per gene in each condition necessary to test this gene\n",
    "def test_gene(gene_df, conditions, col, id_col, rc=10):\n",
    "    \n",
    "    gene_df = gene_df.pivot(index=col, columns=id_col, values='counts')\n",
    "    gene_df = gene_df.transpose()\n",
    "    \n",
    "    groups = gene_df.columns.tolist()\n",
    "    gene_df['total_counts'] = gene_df[groups].sum(axis=1)\n",
    "    gene_df.sort_values(by='total_counts', ascending=False, inplace=True)\n",
    "\n",
    "    if len(gene_df.index) > 11:\n",
    "        gene_df.reset_index(inplace=True)\n",
    "\n",
    "        beep = gene_df.iloc[10:].sum()\n",
    "        beep[id_col] = 'all_other'\n",
    "        beep.index.name = None  \n",
    "        beep = pd.DataFrame(beep).transpose()\n",
    "\n",
    "        gene_df = gene_df.iloc[:10]\n",
    "        gene_df = pd.concat([gene_df, beep])  \n",
    "        \n",
    "    # limit to just isoforms with > 0 expression in both conditions\n",
    "    cond1 = conditions[0]\n",
    "    cond2 = conditions[1]\n",
    "    gene_df = gene_df.loc[(gene_df[cond1]>0)&(gene_df[cond2]>0)]\n",
    "    \n",
    "    # does this gene reach the desired read count threshold?\n",
    "    for cond in conditions:\n",
    "        if gene_df[cond].sum() < rc:\n",
    "            return np.nan, np.nan\n",
    "    \n",
    "    # only do the rest if there's nothing left\n",
    "    if gene_df.empty:\n",
    "        return np.nan, np.nan\n",
    "    \n",
    "    # calculate the percent of each sample each TSS accounts for\n",
    "    cond_pis = []\n",
    "    for cond in conditions:\n",
    "        total_col = '{}_total'.format(cond)\n",
    "        pi_col = '{}_pi'.format(cond)\n",
    "        total_count = gene_df[cond].sum()\n",
    "\n",
    "        cond_pis.append(pi_col)\n",
    "\n",
    "        gene_df[total_col] = total_count\n",
    "        gene_df[pi_col] = (gene_df[cond]/gene_df[total_col])*100\n",
    "        \n",
    "    # compute isoform-level and gene-level delta pis\n",
    "    gene_df['dpi'] = gene_df[cond_pis[0]] - gene_df[cond_pis[1]]\n",
    "    gene_df['abs_dpi'] = gene_df.dpi.abs()\n",
    "    gene_dpi = gene_df.iloc[:2].abs_dpi.sum()    \n",
    "    \n",
    "    # chi squared test \n",
    "    chi_table = gene_df[conditions].to_numpy()\n",
    "    chi2, p, dof, exp = st.chi2_contingency(chi_table)\n",
    "    \n",
    "    return p, gene_dpi\n",
    "\n",
    "def filter_die_results(df, p, dpi):\n",
    "    df = df.loc[(df.adj_p_val<=p)&(df.dpi>=dpi)]\n",
    "    return df\n",
    "\n",
    "# adata: adata with TSS or iso expression \n",
    "# conditions: len 2 list of strings of conditions to compare\n",
    "# col: string, which column the condition labels are in\n",
    "# how: 'tss' or 'iso'\n",
    "def get_die(adata, conditions, how='tss', rc=15):\n",
    "    \n",
    "    if how == 'tss':\n",
    "        id_col = 'tss_id'\n",
    "    elif how == 'iso':\n",
    "        id_col = 'transcript_id'\n",
    "    \n",
    "    # make df that we can groupby\n",
    "    col = 'condition'\n",
    "    colnames = adata.var[id_col].tolist()\n",
    "    print(len(colnames))\n",
    "    print(len(set(colnames)))\n",
    "    rownames = adata.obs.dataset.tolist()    \n",
    "    raw = adata.X\n",
    "    df = pd.DataFrame(data=raw, index=rownames, columns=colnames)\n",
    "    df.reset_index(inplace=True)\n",
    "    df.rename({'index':'dataset'}, axis=1, inplace=True)\n",
    "    samp = adata.obs[['dataset', col]]\n",
    "    df = df.merge(samp, how='left', on='dataset')\n",
    "    \n",
    "    # limit to only the samples that we want in this condition\n",
    "    df[col] = df[col].astype('str')\n",
    "    df = df.loc[df[col].isin(conditions)]\n",
    "        \n",
    "    # groupby sample type and sum over gen\n",
    "    df.drop('dataset', axis=1, inplace=True)\n",
    "    df = df.groupby(col).sum().reset_index()\n",
    "    \n",
    "    # melty boi\n",
    "    tss_cols = df.columns.tolist()[1:]\n",
    "    print(tss_cols[:5])\n",
    "    print(df.head())\n",
    "    df = df.melt(id_vars=col, value_vars=tss_cols)\n",
    "    \n",
    "    # rename some cols\n",
    "    df.rename({'variable':id_col,'value':'counts'}, axis=1, inplace=True)\n",
    "    \n",
    "    # merge with gene names\n",
    "    df = df.merge(adata.var, how='left', on=id_col)\n",
    "    \n",
    "    # get total number of tss or iso / gene\n",
    "    bop = df[['gene_name', id_col]].groupby('gene_name').count().reset_index()\n",
    "    \n",
    "    # construct tables for each gene and test!\n",
    "    gene_names = df.gene_name.unique().tolist()\n",
    "    gene_de_df = pd.DataFrame(index=gene_names, columns=['p_val', 'dpi'], data=[[np.nan for i in range(2)] for j in range(len(gene_names))])\n",
    "    for gene in gene_names:\n",
    "        gene_df = df.loc[df.gene_name==gene]\n",
    "        p, dpi = test_gene(gene_df, conditions, col, id_col, rc=rc)\n",
    "        gene_de_df.loc[gene, 'p_val'] = p\n",
    "        gene_de_df.loc[gene, 'dpi'] = dpi\n",
    "        \n",
    "    # correct p values \n",
    "    gene_de_df.dropna(axis=0, inplace=True)\n",
    "    p_vals = gene_de_df.p_val.tolist()\n",
    "    _, adj_p_vals, _, _ = multipletests(p_vals, method='fdr_bh')\n",
    "    gene_de_df['adj_p_val'] = adj_p_vals\n",
    "    \n",
    "        \n",
    "    return gene_de_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "id": "tough-transmission",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py:4303: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n",
      "Transforming to str index.\n",
      "Transforming to str index.\n"
     ]
    }
   ],
   "source": [
    "ab_file = 'talon/pgp1_tss_talon_abundance.tsv'\n",
    "cond_map = {'Astrocytes': ['astro_1', 'astro_2'], \\\n",
    "            'Excitatory neurons': ['excite_neuron_1', 'excite_neuron_2'], \\\n",
    "            'PGP1': ['pgp1_1', 'pgp1_2']}\n",
    "\n",
    "df = pd.read_csv(ab_file, sep='\\t')\n",
    "adata = make_adata(df, cond_map, how='tss')\n",
    "\n",
    "# do one test for each pair of conditions\n",
    "tested = []\n",
    "conditions = ['Astrocytes', 'Excitatory neurons', 'PGP1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "id": "killing-conducting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20997\n",
      "18064\n",
      "['A1BG_1', 'A1BG-AS1_1', 'AC012313.2_1', 'A2M_1', 'A4GALT_1']\n",
      "            condition  A1BG_1  A1BG-AS1_1  AC012313.2_1  A2M_1  A4GALT_1  \\\n",
      "0          Astrocytes    11.0        11.0          11.0    3.0       1.0   \n",
      "1  Excitatory neurons    29.0        49.0          49.0   40.0      34.0   \n",
      "\n",
      "   AAAS_1  AACS_1  AADAT_1  AADAT_2  ...  ZSWIM7_2  ZSWIM8_1  ZUP1_1  ZW10_1  \\\n",
      "0    22.0    54.0      1.0      8.0  ...       3.0       2.0    32.0    20.0   \n",
      "1    49.0    28.0      0.0      0.0  ...      25.0       0.0    35.0     9.0   \n",
      "\n",
      "   ZWILCH_1  ZWINT_1  ZXDC_1  ZYG11B_1  ZYX_1  ZZEF1_1  \n",
      "0      42.0     18.0    16.0      12.0  543.0      3.0  \n",
      "1      22.0     50.0     2.0       8.0   56.0      5.0  \n",
      "\n",
      "[2 rows x 20998 columns]\n"
     ]
    },
    {
     "ename": "InvalidIndexError",
     "evalue": "Reindexing only valid with uniquely valued Index objects",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mInvalidIndexError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-310-3dd7bfbc634c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0mtested\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mc2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_die\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m         \u001b[0mfname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'{}_{}_{}_die.tsv'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-308-fc084f98b122>\u001b[0m in \u001b[0;36mget_die\u001b[0;34m(adata, conditions, how, rc)\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtss_cols\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    150\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 151\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmelt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mid_vars\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue_vars\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtss_cols\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    152\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0;31m# rename some cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mmelt\u001b[0;34m(self, id_vars, value_vars, var_name, value_name, col_level, ignore_index)\u001b[0m\n\u001b[1;32m   7166\u001b[0m             \u001b[0mvalue_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalue_name\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7167\u001b[0m             \u001b[0mcol_level\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcol_level\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7168\u001b[0;31m             \u001b[0mignore_index\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mignore_index\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7169\u001b[0m         )\n\u001b[1;32m   7170\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/reshape/melt.py\u001b[0m in \u001b[0;36mmelt\u001b[0;34m(frame, id_vars, value_vars, var_name, value_name, col_level, ignore_index)\u001b[0m\n\u001b[1;32m     93\u001b[0m             )\n\u001b[1;32m     94\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 95\u001b[0;31m             \u001b[0midx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mframe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mid_vars\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mvalue_vars\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     96\u001b[0m         \u001b[0mframe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mframe\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     97\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_indexer\u001b[0;34m(self, target, method, limit, tolerance)\u001b[0m\n\u001b[1;32m   2981\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2982\u001b[0m             raise InvalidIndexError(\n\u001b[0;32m-> 2983\u001b[0;31m                 \u001b[0;34m\"Reindexing only valid with uniquely valued Index objects\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2984\u001b[0m             )\n\u001b[1;32m   2985\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mInvalidIndexError\u001b[0m: Reindexing only valid with uniquely valued Index objects"
     ]
    }
   ],
   "source": [
    "how = 'tss'\n",
    "for c1 in conditions:\n",
    "    for c2 in conditions: \n",
    "        if (c1, c2) in tested or c1 == c2 or (c2, c1) in tested:\n",
    "            continue\n",
    "        else:\n",
    "            tested.append((c1,c2))\n",
    "        df = get_die(adata, [c1, c2], how=how, rc=10)\n",
    "        fname = '{}_{}_{}_die.tsv'.format(c1, c2, how)\n",
    "        df.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lesser-enhancement",
   "metadata": {},
   "source": [
    "### Isoform switching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "id": "nuclear-canvas",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/frame.py:4303: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n",
      "Transforming to str index.\n",
      "Transforming to str index.\n"
     ]
    }
   ],
   "source": [
    "pass_list = 'talon/pgp1_pass_list.csv'\n",
    "ab_file = 'talon/pgp1_talon_abundance.tsv'\n",
    "cond_map = {'Astrocytes': ['astro_1', 'astro_2'], \\\n",
    "            'Excitatory neurons': ['excite_neuron_1', 'excite_neuron_2'], \\\n",
    "            'PGP1': ['pgp1_1', 'pgp1_2']}\n",
    "\n",
    "df = pd.read_csv(ab_file, sep='\\t')\n",
    "adata = make_adata(df, cond_map, pass_list=pass_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 305,
   "id": "departmental-technique",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FAM138A-201</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FAM138A-202</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CICP27-201</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>ENST00000442987.3</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AL627309.6-201</td>\n",
       "      <td>AL627309.6</td>\n",
       "      <td>ENST00000494149.2</td>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_transcript_name    gene_name annot_transcript_id      annot_gene_id  \\\n",
       "0       MIR1302-2HG-201  MIR1302-2HG   ENST00000469289.1  ENSG00000243485.5   \n",
       "1           FAM138A-201      FAM138A   ENST00000417324.1  ENSG00000237613.2   \n",
       "2           FAM138A-202      FAM138A   ENST00000461467.1  ENSG00000237613.2   \n",
       "3            CICP27-201       CICP27   ENST00000442987.3  ENSG00000233750.3   \n",
       "4        AL627309.6-201   AL627309.6   ENST00000494149.2  ENSG00000268903.1   \n",
       "\n",
       "   gene_id  transcript_id transcript_novelty ISM_subtype  \n",
       "0        4              6              Known        None  \n",
       "1        6              8              Known        None  \n",
       "2        6              9              Known        None  \n",
       "3       12             21              Known        None  \n",
       "4       13             22              Known        None  "
      ]
     },
     "execution_count": 305,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "wireless-update",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[6, 8, 9, 21, 22]\n",
      "            condition    6    8    9   21   22   32  132  136  138  ...  \\\n",
      "0          Astrocytes  0.0  0.0  0.0  1.0  1.0  7.0  1.0  4.0  0.0  ...   \n",
      "1  Excitatory neurons  1.0  0.0  0.0  0.0  3.0  7.0  2.0  3.0  5.0  ...   \n",
      "\n",
      "   497872  497898  498029  498062  498451  498466  499084  499092  499187  \\\n",
      "0     4.0     1.0    20.0    12.0     0.0    11.0     0.0     0.0     9.0   \n",
      "1    25.0    20.0     1.0     8.0     9.0     2.0    10.0    20.0   123.0   \n",
      "\n",
      "   499325  \n",
      "0    19.0  \n",
      "1     7.0  \n",
      "\n",
      "[2 rows x 40232 columns]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-304-fe190812d24b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m             \u001b[0mtested\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mc2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_die\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0madata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m         \u001b[0mfname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'{}_{}_{}_die.tsv'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mc1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m         \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-301-a220d400dfe8>\u001b[0m in \u001b[0;36mget_die\u001b[0;34m(adata, conditions, how, rc)\u001b[0m\n\u001b[1;32m    162\u001b[0m     \u001b[0mgene_de_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgene_names\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'p_val'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'dpi'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnan\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_names\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    163\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_names\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 164\u001b[0;31m         \u001b[0mgene_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgene_name\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    165\u001b[0m         \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdpi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtest_gene\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconditions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid_col\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    166\u001b[0m         \u001b[0mgene_de_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'p_val'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    877\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    878\u001b[0m             \u001b[0mmaybe_callable\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 879\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmaybe_callable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    880\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    881\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_is_scalar_access\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mTuple\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1088\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_slice_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1089\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1090\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getbool_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1091\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mis_list_like_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1092\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_getbool_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m    896\u001b[0m         \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_bool_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m         \u001b[0minds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnonzero\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 898\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take_with_is_copy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    899\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    900\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_take_with_is_copy\u001b[0;34m(self, indices, axis)\u001b[0m\n\u001b[1;32m   3358\u001b[0m         \u001b[0mSee\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mdocstring\u001b[0m \u001b[0mof\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfull\u001b[0m \u001b[0mexplanation\u001b[0m \u001b[0mof\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3359\u001b[0m         \"\"\"\n\u001b[0;32m-> 3360\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3361\u001b[0m         \u001b[0;31m# Maybe set copy if we didn't actually change the index.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3362\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, indices, axis, is_copy, **kwargs)\u001b[0m\n\u001b[1;32m   3346\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3347\u001b[0m         new_data = self._mgr.take(\n\u001b[0;32m-> 3348\u001b[0;31m             \u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_block_manager_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3349\u001b[0m         )\n\u001b[1;32m   3350\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__finalize__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"take\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mtake\u001b[0;34m(self, indexer, axis, verify, convert)\u001b[0m\n\u001b[1;32m   1446\u001b[0m         \u001b[0mnew_labels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1447\u001b[0m         return self.reindex_indexer(\n\u001b[0;32m-> 1448\u001b[0;31m             \u001b[0mnew_axis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnew_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_dups\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1449\u001b[0m         )\n\u001b[1;32m   1450\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mreindex_indexer\u001b[0;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy, consolidate)\u001b[0m\n\u001b[1;32m   1290\u001b[0m                     ),\n\u001b[1;32m   1291\u001b[0m                 )\n\u001b[0;32m-> 1292\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mblk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1293\u001b[0m             ]\n\u001b[1;32m   1294\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1290\u001b[0m                     ),\n\u001b[1;32m   1291\u001b[0m                 )\n\u001b[0;32m-> 1292\u001b[0;31m                 \u001b[0;32mfor\u001b[0m \u001b[0mblk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1293\u001b[0m             ]\n\u001b[1;32m   1294\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mtake_nd\u001b[0;34m(self, indexer, axis, new_mgr_locs, fill_value)\u001b[0m\n\u001b[1;32m   1253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1254\u001b[0m         new_values = algos.take_nd(\n\u001b[0;32m-> 1255\u001b[0;31m             \u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallow_fill\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mallow_fill\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfill_value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1256\u001b[0m         )\n\u001b[1;32m   1257\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.7/site-packages/pandas/core/algorithms.py\u001b[0m in \u001b[0;36mtake_nd\u001b[0;34m(arr, indexer, axis, out, fill_value, allow_fill)\u001b[0m\n\u001b[1;32m   1706\u001b[0m         \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmask_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmask_info\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1707\u001b[0m     )\n\u001b[0;32m-> 1708\u001b[0;31m     \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfill_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1709\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1710\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mflip_order\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# do one test for each pair of conditions\n",
    "tested = []\n",
    "conditions = ['Astrocytes', 'Excitatory neurons', 'PGP1']\n",
    "\n",
    "how = 'iso'\n",
    "for c1 in conditions:\n",
    "    for c2 in conditions: \n",
    "        if (c1, c2) in tested or c1 == c2 or (c2, c1) in tested:\n",
    "            continue\n",
    "        else:\n",
    "            tested.append((c1,c2))\n",
    "        df = get_die(adata, [c1, c2], how=how, rc=10)\n",
    "        fname = '{}_{}_{}_die.tsv'.format(c1, c2, how)\n",
    "        df.to_csv(fname, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "social-vocabulary",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000442987.3</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000494149.2</td>\n",
       "      <td>ENSG00000268903.1</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_transcript_id      annot_gene_id  gene_ID  transcript_ID  \\\n",
       "0   ENST00000469289.1  ENSG00000243485.5        4              6   \n",
       "1   ENST00000417324.1  ENSG00000237613.2        6              8   \n",
       "2   ENST00000461467.1  ENSG00000237613.2        6              9   \n",
       "3   ENST00000442987.3  ENSG00000233750.3       12             21   \n",
       "4   ENST00000494149.2  ENSG00000268903.1       13             22   \n",
       "\n",
       "  transcript_novelty ISM_subtype  \n",
       "0              Known        None  \n",
       "1              Known        None  \n",
       "2              Known        None  \n",
       "3              Known        None  \n",
       "4              Known        None  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minimal-annex",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
